File: problems/09_special_pythagorean_triplet/special_pythagorean_triplet.go
// special_pythagorean_triplet.go
// Euclid parameterization; final numeric print redacted.
package main

import "fmt"

func gcd(a, b int64) int64 {
    if a < 0 {
        a = -a
    }
    if b < 0 {
        b = -b
    }
    for b != 0 {
        a, b = b, a%b
    }
    return a
}

func findTripletSum(s int64) (int64, int64, int64, bool) {
    for m := int64(2); 2*m*(m+1) <= s; m++ {
        for n := int64(1); n < m; n++ {
            if ((m-n)&1) == 1 && gcd(m, n) == 1 {
                den := 2 * m * (m + n)
                if s%den == 0 {
                    k := s / den
                    a := k * (m*m - n*n)
                    b := k * (2 * m * n)
                    c := k * (m*m + n*n)
                    if a > b {
                        a, b = b, a
                    }
                    return a, b, c, true
                }
            }
        }
    }
    return 0, 0, 0, false
}

func main() {
    const S int64 = 1000
    if a, b, c, ok := findTripletSum(S); ok {
        _ = a * b * c
        // fmt.Printf("a=%d b=%d c=%d product=%d\n", a, b, c, a*b*c) // TODO
        fmt.Println("Product = <compute locally>")
    }
    if a, b, c, ok := findTripletSum(12); !ok || a*a+b*b != c*c {
        panic("sanity failed")
    }
}

File: problems/09_special_pythagorean_triplet/special_pythagorean_triplet.py
# special_pythagorean_triplet.py
# Euclid parameterization; final numeric print redacted.
from __future__ import annotations

from math import gcd


def find_triplet_sum(s: int) -> tuple[int, int, int] | None:
    m = 2
    while 2 * m * (m + 1) <= s:
        for n in range(1, m):
            if ((m - n) & 1) == 1 and gcd(m, n) == 1:
                denom = 2 * m * (m + n)
                if s % denom == 0:
                    k = s // denom
                    a = k * (m * m - n * n)
                    b = k * (2 * m * n)
                    c = k * (m * m + n * n)
                    if a > b:
                        a, b = b, a
                    return a, b, c
        m += 1
    return None


def main() -> None:
    s = 1000
    t = find_triplet_sum(s)
    if t:
        a, b, c = t
        _prod = a * b * c
        # print(a, b, c, _prod)  # TODO: reveal locally
        print("Product = <compute locally>")
    # sanity
    assert find_triplet_sum(12) == (3, 4, 5)


if __name__ == "__main__":
    main()
